### PR TITLE
Kernel/Memory: Various fixes for anonymous mmaps

### PR DESCRIPTION
**Pull request by brody-qq**
This PR fixes the following problems related to anonymous mmaps:
* After a fork, writes to an anonymous mmap  cause redundant page faults (#24637)
* Writes to a shared anonymous mmap are not visible to other processes that share the mmap (#24638)
* VMObject::try_clone() overcommits pages to m_shared_committed_cow_pages

This PR also adds more test cases for anonymous mmaps.

### CODE REVIEW

**Comment by nico** [comment]
@spholz You know this code a bit, right? Want to look at this? :)

File: Kernel/Memory/Region.cpp [hunk 1]
```diff
@@ -505,7 +506,7 @@ PageFaultResponse Region::handle_zero_fault(size_t page_index_in_region, Physica
     RefPtr<PhysicalRAMPage> new_physical_page;
 
     if (page_in_slot_at_time_of_fault.is_lazy_committed_page()) {
-        new_physical_page = static_cast<AnonymousVMObject&>(*m_vmobject).allocate_committed_page({});
+        new_physical_page = anon_vmobject.allocate_committed_page({});
```
Comment (Hendiadyoin1) [inline Kernel/Memory/Region.cpp:509]: This change and the change below look more like a general refactoring, maybe move them into a separate commit

Comment (brody-qq) [reply Kernel/Memory/Region.cpp:509]: Ok, I moved those changes into a new commit

File: Kernel/Memory/Region.cpp [hunk 2]
```diff
@@ -496,6 +496,7 @@ PageFaultResponse Region::handle_zero_fault(size_t page_index_in_region, Physica
 {
     VERIFY(vmobject().is_anonymous());
 
+    auto& anon_vmobject = static_cast<AnonymousVMObject&>(vmobject());
```
Comment (nico) [inline Kernel/Memory/Region.cpp:499]: nit: `anonymous_vmobject` would be more in line with serenityos naming (https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md "Use full words, except in the rare case where an abbreviation would be more canonical and easier to understand.")

**Review by nico — APPROVED** [review]

File: Kernel/Memory/AnonymousVMObject.cpp [hunk 3]
```diff
@@ -274,10 +274,21 @@ NonnullRefPtr<PhysicalRAMPage> AnonymousVMObject::allocate_committed_page(Badge<
     return m_unused_committed_pages->take_one();
 }
 
+void AnonymousVMObject::reset_cow_map()
+{
+    for (size_t i = 0; i < page_count(); ++i) {
+        auto& page = physical_pages()[i];
+        bool should_cow = !page->is_shared_zero_page() && !page->is_lazy_committed_page();
```
Comment (spholz) [inline Kernel/Memory/AnonymousVMObject.cpp:281]: `AnonymousVMObject::should_cow` returns true if the physical page is the shared_zero_page or the lazy_committed_page, so I am unsure if we have other code expecting that pages referencing these physical pages are CoW.

**Review by spholz — COMMENTED** [review]
One small question, but otherwise LGTM

File: Kernel/Memory/AnonymousVMObject.cpp (hunk 3 above)
Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:281]: And `AddressSpace::try_allocate_split_region` also sets the CoW bit for all pages where `Region::should_cow` (which uses `AnonymousVMObjet::should_cow) returns true.

Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:281]: Ah oops, initially an AnonymousVMObject will of course be backed by the actual committed pages or references to either the lazy committed page or the shared zero page.
And those don't have a cow map allocated, so none of the pages are cow.

But `AddressSpace::try_allocate_split_region` should I think still probably not set the cow bit for shared zero pages and lazy committed pages.

Comment (brody-qq) [reply Kernel/Memory/AnonymousVMObject.cpp:281]: Thanks for the feedback, I completely missed the `set_should_cow` call in `AddressSpace::try_allocate_split_region`.

After taking a closer look at `AddressSpace::try_allocate_split_region`, I think the best approach is to just remove the loop that calls `set_should_cow`.
Here is the diff:
```
diff --git a/Kernel/Memory/AddressSpace.cpp b/Kernel/Memory/AddressSpace.cpp
index 9ea53b88a1..e96406e8b1 100644
--- a/Kernel/Memory/AddressSpace.cpp
+++ b/Kernel/Memory/AddressSpace.cpp
@@ -148,11 +148,6 @@ ErrorOr<Region*> AddressSpace::try_allocate_split_region(Region const& source_re
     new_region->set_syscall_region(source_region.is_syscall_region());
     new_region->set_mmap(source_region.is_mmap(), source_region.mmapped_from_readable(), source_region.mmapped_from_writable());
     new_region->set_stack(source_region.is_stack());
-    size_t page_offset_in_source_region = (offset_in_vmobject - source_region.offset_in_vmobject()) / PAGE_SIZE;
-    for (size_t i = 0; i < new_region->page_count(); ++i) {
-        if (source_region.should_cow(page_offset_in_source_region + i))
-            TRY(new_region->set_should_cow(i, true));
-    }
     TRY(m_region_tree.place_specifically(*new_region, range));
     return new_region.leak_ptr();
 }
```

I think this works because if you look at the following line of code (this line comes before the loop), it sets `new_region` to have the same `m_vmobject` as `source_region`
```
auto new_region = TRY(Region::create_unplaced(
        source_region.vmobject(), offset_in_vmobject, move(region_name), source_region.access(), source_region.is_cacheable() ? Region::Cacheable::Yes : Region::Cacheable::No, source_region.is_shared()));
```

So that `set_should_cow` loop really isn't doing anything, since `source_region` and `new_region` reference the same vmobject and cow map.

Does this change make sense? (I hope I'm not missing something obvious here)

Comment (brody-qq) [reply Kernel/Memory/AnonymousVMObject.cpp:281]: Also ProcFS is the only other place I can find that may be impacted by not marking lazy/zero pages in the cow map.

Particularly this line for setting `cow_pages`:
```
           TRY(region_object.add("cow_pages"sv, region.cow_pages()));
```

So if you go with my changes, then ProcFS will report a smaller number than before for `cow_pages`, since it isn't counting the lazy/zero pages anymore.

I don't know enough about how this `cow_pages` number from ProcFS gets used to know if this is a problem, does anyone have any objections to this?

File: Kernel/Memory/Region.cpp (hunk 2 above)
Comment (brody-qq) [reply Kernel/Memory/Region.cpp]: Ok, I replaced `anon_vmobject` with `anonymous_vmobject`

File: Kernel/Memory/AnonymousVMObject.cpp (hunk 3 above)
Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:274]: Removing this loop from `AddressSpace::try_allocate_split_region` seems fine.
This code was added in f4e7aecec24d1aa910e22a782f169efdfdfd564b. Back then the `m_cow_map` was a member of `Region`:
https://github.com/SerenityOS/serenity/blob/f4e7aecec24d1aa910e22a782f169efdfdfd564b/Kernel/VM/Region.h#L193
So this loop should indeed be unnecessary now, since the `m_cow_map` is now owned by `AnonymousVMObject`s.

It may also be nice if we could have a test (in munmap-multi-region-unmapping.cpp maybe?) for splitting a region that is backed by an `AnonymousVMObject` with pages marked as CoW.

And about procfs potentially showing a different `cow_pages` count: This should be fine, we only use that procfs parameter in the `pmap` utility and the "Memory map" tab in SystemMonitor to display the number of CoW pages.

Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:274]: Yes, I think this new test looks fine.

munmap-multi-region-unmapping.cpp is a legacy non-LibTest test. As you may have noticed, those are skipped by default. I am unsure if it is a good idea to add new tests to legacy test files even if they would be the right place for this test.
Sorry, I should have noticed this earlier (especially based on the filename).

@nico what is your opinion as a maintainer on this? (as you already did a review of this pr earlier)

(We also probably should migrate all of those legacy kernel tests)

Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:274]: I don't think that test triggers the `try_allocate_split_region` code path.
`try_allocate_split_region` seems to get called on `munmap`s that don't cover an entire mapping and on `mmap`s with `MAP_FIXED` that overlap, but your `mmap`s don't seem to overlap.

Also, munmap-multi-region-unmapping.cpp is a legacy non-LibTest test. As you may have noticed, those are skipped by default. I am unsure if it is a good idea to add new tests to legacy test files even if they would be the right place for this test.
Sorry, I should have noticed this earlier (especially based on the filename).

**Comment by Hendiadyoin1** [comment]
Seems like one of the other kernel PRs created some conflicts

File: Kernel/Memory/AnonymousVMObject.cpp (hunk 3 above)
Comment (spholz) [reply Kernel/Memory/AnonymousVMObject.cpp:274]: I also don't think this change needs to have a test (I am not a maintainer). I just thought it might be nice to test this behavior.

Comment (brody-qq) [reply Kernel/Memory/AnonymousVMObject.cpp:274]: Ah ok, I misunderstood your original comment. I put the test case you described in `TestAnonymousMmaps.cpp` and got rid of the test I added to the legacy file.

**Comment by brody-qq** [comment]
> Seems like one of the other kernel PRs created some conflicts

Fixed, it was some new code in `Region.cpp` that didn't impact any functionality of this PR.

**Comment by nico** [comment]
Thanks!

Since you're in this part of the code: If you're looking for things to do, it'd be nice if `posix_spawn()` could become a syscall instead of being implemented in userspace. We could save the time needed to copy page metadata then. Most processes are launched using `posix_spawn()` in serenity, so that might help with performance.

**Comment by hiperbolt** [comment]
> Thanks!
> 
> Since you're in this part of the code: If you're looking for things to do, it'd be nice if `posix_spawn()` could become a syscall instead of being implemented in userspace. We could save the time needed to copy page metadata then. Most processes are launched using `posix_spawn()` in serenity, so that might help with performance.

I appologise if commenting on a closed PR is the incorrect form to ask: but is this still something that would be interesting to do?
From what I can tell it is still implemented as a library function and not a system call. Could I start work on a PR to make it a system call?

**Comment by nico** [comment]
That'd be great!

**Comment by ProfessorLongBeard** [comment]
Look ma I'm famous!!
